components/src/main/java/org/folio/rest/workflow/model/resolver/DeserializeAsNodeJsonResolver.java

```
@@ -0,0 +1,107 @@
+    CLASSES.put("ConnectTo", ConnectTo.class);
+    CLASSES.put("DatabaseConnectionTask", DatabaseConnectionTask.class);
+    CLASSES.put("DatabaseDisconnectTask", DatabaseDisconnectTask.class);
+    CLASSES.put("DatabaseQueryTask", DatabaseQueryTask.class);
+    CLASSES.put("DirectoryTask", DirectoryTask.class);
+    CLASSES.put("EmailTask", EmailTask.class);
+    CLASSES.put("EndEvent", EndEvent.class);
+    CLASSES.put("EventSubprocess", EventSubprocess.class);
+    CLASSES.put("ExclusiveGateway", ExclusiveGateway.class);
+    CLASSES.put("FileTask", FileTask.class);
+    CLASSES.put("FtpTask", FtpTask.class);
+    CLASSES.put("InclusiveGateway", InclusiveGateway.class);
+    CLASSES.put("InputTask", InputTask.class);
+    CLASSES.put("MoveToLastGateway", MoveToLastGateway.class);
+    CLASSES.put("MoveToNode", MoveToNode.class);
+    CLASSES.put("Node", Node.class);
+    CLASSES.put("ParallelGateway", ParallelGateway.class);
+    CLASSES.put("ProcessorTask", ProcessorTask.class);
+    CLASSES.put("ReceiveTask", ReceiveTask.class);
+    CLASSES.put("RequestTask", RequestTask.class);
+    CLASSES.put("ScriptTask", ScriptTask.class);
+    CLASSES.put("StartEvent", StartEvent.class);
+    CLASSES.put("Subprocess", Subprocess.class);
+  }}
+
+  private static final Logger logger = LoggerFactory.getLogger(DeserializeAsNodeJsonResolver.class);
+
+  @Override
+  public String idFromValue(Object value) {
+    return value.getClass().getSimpleName();
+  }
+
+  @Override
+  public String idFromValueAndType(Object value, Class<?> suggestedType) {
+    return idFromValue(value);
+  }
+
+  @Override
+  public Id getMechanism() {
+    return Id.SIMPLE_NAME;
+  }
+
+  @Override
+  public JavaType typeFromId(DatabindContext context, String id) {
+    if (CLASSES.containsKey(id)) {
+      return context.getTypeFactory().constructType(CLASSES.get(id));
+    }
+
+    logger.debug("Unknown or Invalid Workflow Model class " + id + ".");
+    return null;
```

Are we sure it would not be better to throw an exception to avoid undetected errors when debug is not on?

---

This is not an error to have an exception.

The NULL return is the intended, designed, and expected behavior.
Example: https://github.com/FasterXML/jackson-databind/blob/bb5b3cc63a5a5594e3fe0d4fd105acffaa09b30a/src/main/java/com/fasterxml/jackson/databind/jsontype/impl/SimpleNameIdResolver.java#L120

---

That is not the same method for the reference implementation `SimpleNameIdResolver`.

https://github.com/FasterXML/jackson-databind/blob/2.19/src/main/java/com/fasterxml/jackson/databind/jsontype/impl/SimpleNameIdResolver.java#L162

https://github.com/FasterXML/jackson-databind/blob/2.19/src/main/java/com/fasterxml/jackson/databind/jsontype/impl/SimpleNameIdResolver.java#L171C16-L171C25

Nevertheless, if the `idToType` map passed into the constructor does not have the key it will return null. But not explicitly and cannot be inferred it is by design but that would be by misuse of the reference implementation.

Here is where the response is null checked. The rabbit hole continues from their chasing null pointers. Why not just throw an exception and handle it without concern for what may change with FasterXML at a later point?

https://github.com/FasterXML/jackson-databind/blob/2.19/src/main/java/com/fasterxml/jackson/databind/jsontype/impl/TypeDeserializerBase.java#L159

---

> That is not the same method for the reference implementation SimpleNameIdResolver.

It shows that they are designing this around expecting NULL.

I don't want to re-implement handling that is already being done elsewhere.
They are clearly expecting and handling NULL as you linked here:
- https://github.com/FasterXML/jackson-databind/blob/bb5b3cc63a5a5594e3fe0d4fd105acffaa09b30a/src/main/java/com/fasterxml/jackson/databind/jsontype/impl/TypeDeserializerBase.java#L159

And here is how they show that the allow NULL returns:
- https://github.com/FasterXML/jackson-databind/blob/2.19/src/main/java/com/fasterxml/jackson/databind/deser/std/NullifyingDeserializer.java

And:
- https://github.com/FasterXML/jackson-databind/blob/bb5b3cc63a5a5594e3fe0d4fd105acffaa09b30a/src/main/java/com/fasterxml/jackson/databind/deser/std/NullifyingDeserializer.java#L51

And:
- https://github.com/FasterXML/jackson-databind/blob/bb5b3cc63a5a5594e3fe0d4fd105acffaa09b30a/src/main/java/com/fasterxml/jackson/databind/deser/std/NullifyingDeserializer.java#L66

We suffer a lot of maintainability problems, upgrading problems, and migration problems from re-implementing or re-handling what is already being handled.

Rather than re-implement the wheel of exception handling, let it properly handle its own errors as it expects.

Here is what happens when I change the `deserializeAs` to something like `Workflow` rather than `StartEvent`:
```
 at [Source: (org.springframework.util.StreamUtils$NonClosingInputStream); line: 1, column: 165]
com.fasterxml.jackson.databind.exc.InvalidTypeIdException: Could not resolve type id 'Workflow' as a subtype of `org.folio.rest.workflow.model.Node`: type ids are not statically known
```

The exception that is already thrown by Jackson seems perfectly fine to me.
I do not want to re-implement this.